src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala

```diff
@@ -115,6 +115,35 @@ trait ClusterComponent extends LeoComponent {
       }
     }
 
+    def listByLabels(labelMap: Map[String, String]): DBIO[Seq[Cluster]] = {
+      val query = if (labelMap.isEmpty) {
+        clusterQueryWithLabels
+      } else {
+        // The trick is to find all clusters that have _at least_ all the labels in labelMap.
+        // In other words, for a given cluster, the labels provided in the query string must be
+        // a subset of its labels in the DB. The following SQL achieves this:
+        //
+        // select c.*, l.*
+        // from cluster c
+        // left join label l on l.clusterId = c.id
+        // where (
+        //   select count(*) from label
+        //   where clusterId = c.id and (key, value) in ${labelMap}
+        // ) = ${labelMap.size}
+        //
+        clusterQueryWithLabels.filter { case (cluster, _) =>
+          labelQuery.filter { _.clusterId === cluster.id }
+            // The following confusing line is equivalent to the much simpler:
+            // .filter { lbl => (lbl.key, lbl.value) inSetBind labelMap.toSet }
+            // Unfortunately slick doesn't support inSet/inSetBind for tuples.
+            // https://github.com/slick/slick/issues/517
+            .filter { lbl => labelMap.map { case (k, v) => lbl.key === k && lbl.value === v }.reduce(_ || _) }
```

Comment:
Can you explain in a quick sentence what `.reduce(_ || _)` does?

Reply:
`reduce` is a like a `fold` without a zero value (the first element is used as the zero instead). In this case, it combines a list of booleans into one boolean using `||`.

```
scala> List(true, false, false).reduce(_ || _)
res0: Boolean = true

scala> List(true, false, false).reduce(_ && _)
res1: Boolean = false
```

This is basically to generate a SQL fragment like this:
```
where (key = "key1" and value = "value1") or (key = "key2" and value = "value2") or ...
```
which is equivalent to (if slick supported this):
```
where (key, value) in (("key1", "value1"), ("key2", "value2"), ...)
```

Reply:
Cool! Thank you for the explanation! :)